File: go/otp_test.go
package realmid

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"testing"
)

func TestOTP_Issue(t *testing.T) {
	var gotBody map[string]any
	var gotBearer, gotOBO string
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/otp/issue": func(w http.ResponseWriter, r *http.Request) {
			gotBearer = r.Header.Get("Authorization")
			gotOBO = r.Header.Get("X-On-Behalf-Of-User")
			buf, _ := io.ReadAll(r.Body)
			_ = json.Unmarshal(buf, &gotBody)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id":          "otp-1",
				"value":       "123456",
				"expires_at":  "2026-05-08T12:00:00Z",
				"purpose":     "delivery",
				"subject_ref": "booking:X",
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	resp, err := r.OTP.Issue(context.Background(), IssueRequest{
		SubjectRef: "booking:X",
		Purpose:    "delivery",
		UserID:     "manager-A",
	})
	if err != nil {
		t.Fatalf("Issue: %v", err)
	}
	if resp.Value != "123456" || resp.ID != "otp-1" {
		t.Errorf("resp: %+v", resp)
	}
	if gotBody["subject_ref"] != "booking:X" || gotBody["purpose"] != "delivery" {
		t.Errorf("body: %+v", gotBody)
	}
	if gotBearer != "Bearer ptok" {
		t.Errorf("bearer: %q", gotBearer)
	}
	if gotOBO != "manager-A" {
		t.Errorf("on-behalf-of: %q", gotOBO)
	}
	if resp.ExpiresAt.IsZero() {
		t.Error("ExpiresAt not parsed")
	}
}

func TestOTP_Verify_Success(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/otp/verify": func(w http.ResponseWriter, r *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"otp_id":         "otp-1",
				"issuer_user_id": "manager-A",
				"issued_at":      "2026-05-08T11:00:00Z",
				"subject_ref":    "booking:X",
				"purpose":        "delivery",
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	resp, err := r.OTP.Verify(context.Background(), VerifyRequest{
		SubjectRef: "booking:X", Purpose: "delivery", Presented: "123456",
		UserID: "agent-svc",
	})
	if err != nil {
		t.Fatalf("Verify: %v", err)
	}
	if resp.IssuerUserID != "manager-A" {
		t.Errorf("issuer = %q", resp.IssuerUserID)
	}
}

func TestOTP_Verify_Invalid(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/otp/verify": func(w http.ResponseWriter, _ *http.Request) {
			w.WriteHeader(http.StatusUnauthorized)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"error": map[string]any{"code": "invalid_otp", "message": "invalid OTP"},
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	_, err := r.OTP.Verify(context.Background(), VerifyRequest{
		SubjectRef: "booking:X", Purpose: "delivery", Presented: "wrong",
		UserBearer: "user-jwt",
	})
	var re *RealmError
	if !errors.As(err, &re) {
		t.Fatalf("expected RealmError, got %v", err)
	}
	if re.Code != "invalid_otp" {
		t.Errorf("code = %q", re.Code)
	}
}

func TestOTP_View_IssuerOnly(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/otp/otp-1": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id":             "otp-1",
				"value":          "654321",
				"expires_at":     "2026-05-08T12:00:00Z",
				"purpose":        "login",
				"subject_ref":    "user:bob",
				"issuer_user_id": "manager-A",
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	resp, err := r.OTP.View(context.Background(), "otp-1", ViewOptions{UserID: "manager-A"})
	if err != nil {
		t.Fatalf("View: %v", err)
	}
	if resp.Value != "654321" || resp.IssuerUserID != "manager-A" {
		t.Errorf("resp: %+v", resp)
	}
}

File: ts/src/otp.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

interface Recorded {
  url: string;
  method: string;
  headers: Headers;
  body: unknown;
}

function recorder(handlers: Array<(rec: Recorded) => Response | Promise<Response>>): {
  fetch: typeof fetch;
  calls: Recorded[];
} {
  const calls: Recorded[] = [];
  let i = 0;
  const fetch = (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const headers = new Headers(init?.headers);
    let body: unknown;
    if (init?.body) {
      try { body = JSON.parse(String(init.body)); } catch { body = init.body; }
    }
    const rec: Recorded = { url, method: init?.method ?? "GET", headers, body };
    calls.push(rec);

    if (url.endsWith("/auth/login")) {
      return new Response(JSON.stringify({
        platform_token: "pt_test",
        expires_in: 300,
        realm_id: "01HREALM",
      }), { status: 200, headers: { "content-type": "application/json" } });
    }
    const h = handlers[i++] ?? handlers[handlers.length - 1]!;
    return h(rec);
  }) as typeof fetch;
  return { fetch, calls };
}

const REALM_ID = "01HREALM";
const API_KEY = "rk_live_test";

test("otp.issue: posts subject_ref + purpose, returns response", async () => {
  const { fetch, calls } = recorder([
    () => new Response(JSON.stringify({
      id: "otp-1",
      value: "123456",
      expires_at: "2026-05-08T12:00:00Z",
      purpose: "delivery",
      subject_ref: "booking:X",
    }), { status: 200, headers: { "content-type": "application/json" } }),
  ]);
  const realm = createRealm({ realmId: REALM_ID, apiKey: API_KEY, baseUrl: "https://auth.test", fetch });
  const out = await realm.otp.issue({
    subjectRef: "booking:X",
    purpose: "delivery",
    userBearer: "user-jwt",
  });
  assert.equal(out.id, "otp-1");
  assert.equal(out.value, "123456");
  assert.equal(out.expiresAt, "2026-05-08T12:00:00Z");

  const issueCall = calls.find((c) => c.url.endsWith("/auth/otp/issue"))!;
  assert.equal((issueCall.body as Record<string, unknown>).subject_ref, "booking:X");
  assert.equal(issueCall.headers.get("authorization"), "Bearer user-jwt");
});

test("otp.verify: success returns issuer attribution", async () => {
  const { fetch } = recorder([
    () => new Response(JSON.stringify({
      otp_id: "otp-1",
      issuer_user_id: "manager-A",
      issued_at: "2026-05-08T11:00:00Z",
      subject_ref: "booking:X",
      purpose: "delivery",
    }), { status: 200, headers: { "content-type": "application/json" } }),
  ]);
  const realm = createRealm({ realmId: REALM_ID, apiKey: API_KEY, baseUrl: "https://auth.test", fetch });
  const out = await realm.otp.verify({
    subjectRef: "booking:X",
    purpose: "delivery",
    presented: "123456",
    userBearer: "agent-svc-jwt",
  });
  assert.equal(out.otpId, "otp-1");
  assert.equal(out.issuerUserId, "manager-A");
});

test("otp.verify: invalid surfaces invalid_otp", async () => {
  const { fetch } = recorder([
    () => new Response(JSON.stringify({
      error: { code: "invalid_otp", message: "invalid OTP" },
    }), { status: 401, headers: { "content-type": "application/json" } }),
  ]);
  const realm = createRealm({ realmId: REALM_ID, apiKey: API_KEY, baseUrl: "https://auth.test", fetch });
  await assert.rejects(
    () => realm.otp.verify({
      subjectRef: "booking:X", purpose: "delivery", presented: "wrong",
      userBearer: "agent-svc-jwt",
    }),
    (err) => err instanceof RealmError && err.code === "invalid_otp",
  );
});

test("otp.view: returns issuerUserId", async () => {
  const { fetch } = recorder([
    () => new Response(JSON.stringify({
      id: "otp-1",
      value: "654321",
      expires_at: "2026-05-08T12:00:00Z",
      purpose: "login",
      subject_ref: "user:bob",
      issuer_user_id: "manager-A",
    }), { status: 200, headers: { "content-type": "application/json" } }),
  ]);
  const realm = createRealm({ realmId: REALM_ID, apiKey: API_KEY, baseUrl: "https://auth.test", fetch });
  const out = await realm.otp.view("otp-1", { userBearer: "manager-jwt" });
  assert.equal(out.value, "654321");
  assert.equal(out.issuerUserId, "manager-A");
});
